File: Python_Ver/main_real.py
import os
import numpy as np
from scipy.io import loadmat
from utils import *
from training import *

# Ensure the results directory exists
if not os.path.exists('./Python_Ver/results'):
    os.makedirs('./Python_Ver/results')

# List of dataset names
DataNameList = ["./Python_Ver/data/rat01.mat", "Python_Ver/data/rat02.mat"]

for DataName in DataNameList:
    data = data_setup(DataName)  # Load data
    opt = opt_Setup_real(data)  # Set options
    
    results_dir = f'./Python_Ver/results/{opt["DataIndex"]}/'
    if not os.path.exists(results_dir):
        os.makedirs(results_dir)
    
    np.random.seed(0)  # Set random seed to default
    
    # PreProcess: get mPFC ensemble for every time bin
    inputEnsemble, M1_truth, Actions, Trials, opt = PreProcess(
        data["mPFC"], data["M1"], data["segment"], data["trialNo"], opt["decay_parameter"], opt
    )
    for sch in ["RL", "Sup"]:
        for testFold in range(opt["foldNum"]):  # 5-fold cross-validation
            opt["testFold"] = testFold
            trainFolds = list(range(opt["foldNum"]))
            trainFolds.remove(testFold)
            opt["trainTrials"] = np.concatenate([opt["folds"][i] for i in trainFolds])
            opt["NumberOfTrainTrials"] = len(opt["trainTrials"])
            opt["testTrials"] = opt["folds"][testFold] 
            opt["NumberOfTestTrials"] = len(opt["testTrials"])
            
            print(f'----{opt["DataIndex"]} {sch} fold {opt["testFold"]} Train start----')
            
            for _ in range(opt["ReTrainTimes"]):
                if sch == 'RL':
                    RLPP(data, opt, inputEnsemble, M1_truth, Actions, Trials)
                else:   
                    SupervisedLearning(data, opt, inputEnsemble, M1_truth, Actions, Trials)
            
            # Evaluate results
            paths = [f for f in os.listdir(results_dir) if f.startswith(f'{sch}_{opt["testFold"]}_') and f.endswith('.mat')]
            SucRate = []
            testSuc = []
            
            for path in paths:
                file_path = os.path.join(results_dir, path)
                data_dict = loadmat(file_path)
                SucRate.append(data_dict["MaxReward"].item())
                testSuc.append(data_dict["testSucRate"].item())
            
            best_idx = np.argmax(SucRate)
            best_path = paths[best_idx]
            source = os.path.join(results_dir, best_path)
            destination = f'./Python_Ver/results/{data["DataIndex"]}_{sch}_{opt["testFold"]}.mat'
            
            os.replace(source, destination)
            print(f'{destination} testSuc: {testSuc[best_idx]}')

File: main_real.m
clc;clear;close all;
addpath model/ utils/ decoding/ training/

if ~exist('./results', 'dir')
    mkdir('./results');
end

%% model training
DataNameList = ["data/rat01.mat", "data/rat02.mat"];

for dataIdx = 1:2
    DataName = DataNameList(dataIdx);
    data = data_setup(DataName); % load data
    opt = opt_Setup_real(data); % set options

    if ~exist(['./results/', opt.DataIndex,'/'], 'dir')
        mkdir(['./results/', opt.DataIndex,'/']);
    end
    
    rng('default');
    % PreProcess: get mPFC ensemble for every time bin
    [inputEnsemble,M1_truth,Actions,Trials,opt] = PreProcess( ...
        data.mPFC,data.M1,data.segment,data.trialNo,opt.decay_parameter,opt);
    for RLorSL = [0 1]
        if RLorSL==0
            sch = 'RL';
        else
            sch = 'Sup';
        end
    
        for testFold = 1:opt.foldNum  % explore testFold in 1:5 for 5-fold cross validation
            opt.testFold = testFold;
            trainFolds = 1:opt.foldNum; trainFolds(testFold) = [];
            opt.trainTrials = cell2mat(opt.folds(trainFolds));
            opt.NumberOfTrainTrials = length(opt.trainTrials);
            opt.testTrials = opt.folds{testFold};
            opt.NumberOfTestTrials = length(opt.testTrials);
        
            disp(['----', opt.DataIndex,' ',sch,' fold ',num2str(opt.testFold),' Train start----']);
            parfor i = 1:opt.ReTrainTimes
                s = rng;
                if (RLorSL==0)
                    RLPP(data,opt,s,inputEnsemble,M1_truth,Actions,Trials);
                else
                    SupervisedLearning(data,opt,s,inputEnsemble,M1_truth,Actions,Trials);
                end
            end

            paths = dir(['results/',data.DataIndex,'/',sch,'_',num2str(opt.testFold),'_','*.mat']);
            SucRate = zeros(1,length(paths));
            testSuc = zeros(1,length(paths));
            for i=1:length(paths)
                load(['results/',data.DataIndex,'/',paths(i).name],"MaxReward","testSucRate")
                SucRate(i) = MaxReward;
                testSuc(i) = testSucRate;
            end
            [best,I] = max(SucRate);
            source = ['results/',data.DataIndex,'/',paths(I).name];
            destination = ['results/', data.DataIndex, '_', sch, '_', num2str(opt.testFold), '.mat'];
            copyfile(source, destination, 'f')
            disp([destination, ' testSuc: ', num2str(testSuc(I))])
        end
    end
end
